README.md
# ApiAxle

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/apiaxle/apiaxle?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

http://apiaxle.com

A free, locally hosted API management solution. A proxy for your api,
statistics for your api & a powerful api of its own.

There are three components which make up the Api Axle system:

## The proxy

    $ npm install apiaxle-proxy

This is the aspect of the system which does the actual proxying. It
sits in front of your API and does the authentication, key checking,
queries per day/second checking. This is the bit you want if you want
anything. More detail on the [main site](http://apiaxle.com).

## The API

    $ npm install apiaxle-api

This is the (optional) API for managing users, keys and
endpoints. Once installed, run it with:

    $ apiaxle-api

## The REPL

    $ npm install apiaxle-repl

A way to administer your ApiAxle installation via a command line. Once
installed, run it with:

    $ apiaxle

You then get a prompt where you can type `help` to find out more.

## The base libs

This is a set of libraries which is required for the above components.

---

# Installation

Check the [main site](http://apiaxle.com) for more detailed
installation instructions.

# Build

* master: [![Build Status](https://secure.travis-ci.org/apiaxle/apiaxle.png?branch=master)](http://travis-ci.org/apiaxle/apiaxle)
* develop: [![Build Status](https://secure.travis-ci.org/apiaxle/apiaxle.png?branch=develop)](http://travis-ci.org/apiaxle/apiaxle)

# Docker

## Docker image
This repository is auto-built and published as
[apiaxle/apiaxle](https://hub.docker.com/r/apiaxle/apiaxle/).

## Dockerfiles
This project uses two `Dockerfile`s, one for production usage named `Dockerfile`,
and a second for development of ApiAxle itself, named `Dockerfile-development`.
Docker Compose is configured to build the development version for you, see
below for examples.

## Environment Variables
Environment variables can be used to configure `NODE_ENV`, `REDIS_HOST`,
`REDIS_PORT`, `API_NAME_REGEX` and `DEBUG_MODE`. Below are defaults for production:

 - `NODE_ENV` = `production`
 - `REDIS_HOST` = `redis`
 - `REDIS_PORT` = `6379`
 - `API_NAME_REGEX` = `^(.+?)\\.api\\.` _(extra backslash required for JSON)_
 - `DEBUG_MODE` = `false`

## Start services
```
docker-compose up -d redis
docker-compose up -d api
docker-compose up -d proxy
```

## Run repl
```
docker-compose run repl
```

## Run tests
```
docker-compose run repl test
```

## Enter container
```
docker-compose run --entrypoint sh repl
```

---

### Example adding an api and key:
```
api acme create endPoint='localhost:8000'
key 1234 create
api acme linkkey 1234
```

### Example curl:
```
curl localhost:3000?api_key=1234 -H 'Host: acme.api.localhost'
```


CHANGELOG.md
changelog
=========

1.14.0
------
- production ready docker
- api name regex

1.13.1
------
- all apis keys

1.13.0
-------
- cascading limits

1.12.38
-------
- error message customization

1.12.36
-------
- support external hit processors for apiaxle-proxy
- add qpm limits

1.12.35
-------
- add batch endpoint to apiaxle-api


setup-apiaxle.bash
#!/usr/bin/env bash

set -e

function which-or-die {
  for command in "${@}"; do
    if ! which "${command}"; then
      echo "The '${command}' program is missing please install and try again!"
      exit 1
    fi
  done
}

function save-excurstion {
  dir="${1}"; shift

  pwd
  pushd "${dir}"
  ${@}
  popd
}

function install-ubuntu-packages {
  for package in "${@}"; do
    if ! dpkg -s "${package}" &>/dev/null; then
      sudo apt-get install "${@}"
      break
    fi
  done
}

which-or-die node npm

# try to install ubuntu pre-reqs
if [[ -f /proc/version ]]; then
  if grep Ubuntu /proc/version &>/dev/null; then
    echo "Detected ubuntu, attempting to install pre-reqs:"
    install-ubuntu-packages "libxml2-dev" "build-essential"
  fi
fi

git clone git://github.com/apiaxle/apiaxle.git

cd apiaxle

# have npm do its thing
for d in base proxy api; do
  save-excurstion "${d}" npm install
done

for d in proxy api; do
  save-excurstion "${d}" npm install ../base
done

npm install -g coffee-script

echo -e "\n\nHopefully you're done. Have a look at the documentation:" >&2
echo    "  http://apiaxle.com/docs/try-it-now/" >&2
echo    "to see how to proceed." >&2


docker-entrypoint.sh
#!/bin/sh

port=3000
app='api'

# app can be passed as first parameter
if [ $# -gt 0 ]; then
  app=$1
fi

# port can be passed as second parameter
if [ $# -gt 1 ]; then
  port=$2
fi

# Set default values if env vars are not present
if [ -z $REDIS_HOST ]; then
  REDIS_HOST=redis
fi
if [ -z $REDIS_PORT ]; then
  REDIS_PORT=6379
fi
if [ -z $DEBUG_MODE ]; then
  DEBUG_MODE=false
fi
if [ -z $API_NAME_REGEX ]; then
  API_NAME_REGEX="^(.+?)\\.api\\."
fi

# Update config file with env values
mkdir -p /etc/apiaxle
CONFIG_FILE=/etc/apiaxle/$NODE_ENV.json
cp /apiaxle-config.json $CONFIG_FILE
sed -i "s/REDIS_HOST/${REDIS_HOST}/" $CONFIG_FILE
sed -i "s/REDIS_PORT/${REDIS_PORT}/" $CONFIG_FILE
sed -i "s/DEBUG_MODE/${DEBUG_MODE}/" $CONFIG_FILE
sed -i "s/API_NAME_REGEX/${API_NAME_REGEX}/" $CONFIG_FILE

if [ $app = 'repl' ]; then
  cd /app/apiaxle/repl
  if [ $NODE_ENV != 'production' ]; then
    # When NODE_ENV isn't 'production', run as coffee for watch and recompile
    coffee --watch /app/apiaxle/base/index.coffee /app/apiaxle/base &
    coffee /app/apiaxle/repl/apiaxle.coffee
  else
    node /app/apiaxle/repl/apiaxle.js
  fi

elif [ $app = 'test' ]; then
  cd /app/apiaxle
  make
  make test

elif [ $app = 'docs' ]; then
  cd /app/apiaxle
  make > /dev/null
  cd /app/apiaxle/api
  bin/generate-docs.coffee

else
  if [ $NODE_ENV != 'production' ]; then
    # When NODE_ENV isn't 'production', run as coffee for watch and recompile
    coffee --watch /app/apiaxle/base/index.coffee /app/apiaxle/base &
    supervisor --force-watch -x coffee -w "/app/apiaxle/$app,/app/apiaxle/base" -- /app/apiaxle/$app/apiaxle-$app.coffee -h 0.0.0.0 -p $port -f 1
  else
    cd /app/apiaxle/$app
    node /app/apiaxle/$app/apiaxle-$app.js -h 0.0.0.0 -p $port -f 1
  fi
fi
